### PR TITLE
[Apple Silicon] Add a custom Metal RMSNorm kernel

### PR DESCRIPTION
**Pull request by SasankYadati**
## What this adds

This adds a custom Metal RMSNorm kernel for the Apple Silicon MLX backend. It uses the same compiled ahead of time build path as the merged RoPE kernel (#22868 on the #23449 infra). The kernel is off by default. You turn it on with an environment variable, and it falls back to `mx.fast.rms_norm` when it is off or when a call is not supported.

## Code

Kernel and build:
- `sgl-kernel/csrc/metal/rms_norm.metal`: the kernel. It computes `y = x * rsqrt(mean(x^2, axis=-1) + eps) * w`. It accumulates in fp32, upcasts the weight to fp32, and casts back to the input dtype on store. It loads four elements at a time when the hidden size is a multiple of 4, and uses a scalar path otherwise.
- `sgl-kernel/csrc/metal/rms_norm.cpp`: the MLX Primitive, the dispatch, and the Python entry point.
- `sgl-kernel/csrc/metal/metal_common.h`: a small shared header for the dtype suffix helper.
- `sgl-kernel/setup_metal.py`: registers the two new source files.
- `sgl-kernel/python/sgl_kernel/metal.py`: the Python wrapper.

Backend integration:
- `python/sglang/srt/hardware_backend/mlx/norm_wrapper.py`: a wrapper that replaces each plain `nn.RMSNorm` module and calls the kernel, plus `patch_model_norms` which installs it. It is gated by `SGLANG_MLX_USE_CUSTOM_RMSNORM`.
- The wrapper is installed in `model_runner.py` right after the attention wrapper.

## Correctness

- `sgl-kernel/tests/test_metal_norm.py`: 99 cases pass. 96 of them compare the kernel to an fp32 reference across the same sweep as `test_norm.py`: batch sizes {1, 19, 99, 989}, hidden sizes {111, 500, 1024, 3072, 3584, 4096, 8192, 16384}, dtypes {f16, bf16, f32}. The other 3 check input validation. Tolerances match `test_norm.py`: 1e-3 for f16, 2e-2 for bf16, 1e-4 for f32. Worst observed max difference: 3.9e-3 for f16, 3.1e-2 for bf16, 1.9e-6 for f32. The hidden size list covers both the vectorized path (multiples of 4) and the scalar path (111, 500). The test is skipped when the Metal extension is not built, so it is a no-op off Apple Silicon.
- End to end on Qwen3-0.6B, checked two ways. Through the server, the greedy `output_ids` are identical with the flag on and off. In a standalone model-level check that swaps only the norm implementation, the argmax token matches at every prompt position. The last-token logits differ by up to 0.34 in absolute value, which is accumulated rounding across 113 low precision norms and does not change any greedy token.

## Performance

Measured with `sgl-kernel/benchmark/bench_metal_rmsnorm.py` against `mx.fast.rms_norm` on an **M4 Pro (48 GB), mlx 0.31.2, macOS 26**. The harness uses a pool of distinct inputs, takes the minimum over 7 repeats, and clears the MLX cache between configs. Numbers are microseconds per call, amortized inside one lazy evaluation. Speedup above 1.0 means the custom kernel is faster.

image

Summary:

1. **Ahead at prefill-like shapes with small hidden sizes:** up to 1.6x at batch 512 with hidden 896 to 1024, and 1.1x to 1.4x at batch 128 with hidden up to 3072 (f16).
2. **At parity at large batch:** 0.96x to 1.08x at batch 2048 across all hidden sizes.
3. **Behind at small batches, which is low concurrency decode:** 0.7x to 0.9x at batch 1 to 32. The absolute gap is small (the calls are 5 to 12 microseconds, launch overhead dominated), but it is consistent.
4. **Slightly behind at medium batch with large hidden:** 0.85x to 0.99x at batch 128 to 512 with hidden 2048 and up. `mx.fast.rms_norm` is notably strong for bf16 here.

The fuller 84-config sweep is below. Given this profile, I would keep the kernel opt-in rather than on by default for now. The standalone op wins in a region, and the case for default-on is potentially the fused add + norm follow up, which removes a memory round trip. An alternative is to route by shape in the wrapper (e.g., small row counts go to `mx.fast.rms_norm`), but that threshold would be tuned on one chip only. 

Open to hear feedback.

<details>
<summary>Full sweep (84 configs, f16 and bf16, batch 1 to 2048, hidden 896 to 8192)</summary>

```
dtype      batch  hidden   metal_us  mxfast_us  speedup
-------------------------------------------------------
float16        1     896       7.62       6.07    0.80x
float16        1    1024       8.24       5.71    0.69x
float16        1    2048       8.29       5.44    0.66x
float16        1    3072       7.60       5.85    0.77x
float16        1    4096       7.35       6.45    0.88x
float16        1    5120       7.05       5.34    0.76x
float16        1    8192       8.07       5.82    0.72x
float16        8     896       7.44       8.08    1.09x
float16        8    1024       7.67       7.96    1.04x
float16        8    2048       7.80       7.73    0.99x
float16        8    3072       7.74       7.51    0.97x
float16        8    4096       8.18       7.48    0.91x
float16        8    5120       8.60       8.30    0.97x
float16        8    8192       9.70       8.77    0.90x
float16       32     896       9.25       9.85    1.06x
float16       32    1024       8.91      10.05    1.13x
float16       32    2048       9.55      10.70    1.12x
float16       32    3072      10.51      10.43    0.99x
float16       32    4096      11.59      11.60    1.00x
float16       32    5120      11.60      14.78    1.27x
float16       32    8192      11.58      10.48    0.90x
float16      128     896       9.46      11.97    1.26x
float16      128    1024      13.01      18.63    1.43x
float16      128    2048      15.82      18.40    1.16x
float16      128    3072      18.33      20.28    1.11x
float16      128    4096      15.82      13.68    0.87x
float16      128    5120      17.92      16.82    0.94x
float16      128    8192      25.17      23.96    0.95x
float16      512     896      14.89      23.78    1.60x
float16      512    1024      15.57      24.79    1.59x
float16      512    2048      24.84      23.48    0.95x
float16      512    3072      33.63      32.05    0.95x
float16      512    4096      43.77      42.21    0.96x
float16      512    5120      54.53      53.92    0.99x
float16      512    8192      88.60      85.83    0.97x
float16     2048     896      37.16      37.40    1.01x
float16     2048    1024      41.32      43.57    1.05x
float16     2048    2048      81.87      78.70    0.96x
float16     2048    3072     126.68     124.84    0.99x
float16     2048    4096     168.30     182.59    1.08x
float16     2048    5120     205.48     212.73    1.04x
float16     2048    8192     337.77     344.80    1.02x
bfloat16       1     896       6.60       4.70    0.71x
bfloat16       1    1024       6.50       4.63    0.71x
bfloat16       1    2048       5.60       3.72    0.66x
bfloat16       1    3072       5.42       3.89    0.72x
bfloat16       1    4096       5.57       3.69    0.66x
bfloat16       1    5120       6.29       4.74    0.75x
bfloat16       1    8192       6.54       4.74    0.73x
bfloat16       8     896       6.25       5.56    0.89x
bfloat16       8    1024       6.65       6.49    0.98x
bfloat16       8    2048       6.82       5.58    0.82x
bfloat16       8    3072       7.11       5.87    0.83x
bfloat16       8    4096       7.19       5.78    0.80x
bfloat16       8    5120       7.24       6.03    0.83x
bfloat16       8    8192       7.20       5.98    0.83x
bfloat16      32     896       7.67       6.27    0.82x
bfloat16      32    1024       7.35       7.02    0.95x
bfloat16      32    2048       7.38       6.24    0.85x
bfloat16      32    3072       8.00       6.42    0.80x
bfloat16      32    4096       7.16       5.99    0.84x
bfloat16      32    5120       7.65       6.27    0.82x
bfloat16      32    8192      10.08       8.18    0.81x
bfloat16     128     896       7.95       8.76    1.10x
bfloat16     128    1024       8.31       8.85    1.06x
bfloat16     128    2048      10.94       9.70    0.89x
bfloat16     128    3072      13.06      11.10    0.85x
bfloat16     128    4096      15.45      13.72    0.89x
bfloat16     128    5120      17.80      15.76    0.88x
bfloat16     128    8192      25.67      22.73    0.89x
bfloat16     512     896      14.90      19.44    1.30x
bfloat16     512    1024      15.37      19.85    1.29x
bfloat16     512    2048      24.11      22.68    0.94x
bfloat16     512    3072      33.48      31.54    0.94x
bfloat16     512    4096      43.58      42.51    0.98x
bfloat16     512    5120      55.05      53.24    0.97x
bfloat16     512    8192      88.94      86.54    0.97x
bfloat16    2048     896      37.14      37.70    1.01x
bfloat16    2048    1024      41.37      43.77    1.06x
bfloat16    2048    2048      83.09      79.58    0.96x
bfloat16    2048    3072     127.03     124.38    0.98x
bfloat16    2048    4096     171.90     184.03    1.07x
bfloat16    2048    5120     213.26     217.09    1.02x
bfloat16    2048    8192     331.82     330.54    1.00x
```

The batch 1, hidden 896, f16 cell reads 0.35x when it is the first config a fresh process runs, which looks like GPU clock ramp up. The value above is from a warm re-run. All other cells are stable across repeated runs.

</details>

## How to test

- Build: `python sgl-kernel/setup_metal.py install` on Apple Silicon with Python 3.11.
- Run the tests: `python -m pytest sgl-kernel/tests/test_metal_norm.py`.
- Enable it in the server with `SGLANG_MLX_USE_CUSTOM_RMSNORM=1` alongside `SGLANG_USE_MLX=1`.

---
### CI States

Latest PR Test (Base): :x: [Run #30717815848](https://github.com/sgl-project/sglang/actions/runs/30717815848)
Latest PR Test (Extra): :x: [Run #30717815724](https://github.com/sgl-project/sglang/actions/runs/30717815724)

### CODE REVIEW

**Review by jlee5814 — COMMENTED** [review]
Please run `pre-commit run --all-files` for the lint failure, and add a `test_norm_patching.py` (mirroring `test_attention_patching.py`) covering patch idempotency, the Gemma rejection, and the unsupported-shape fallback. The kernel's tested but the wrapper isn't.
